### ISSUE TITLE
Polymorphism Support

### ISSUE THREAD

**Issue by ruediste** [vojtechhabarta/typescript-generator#397]
I am looking to add property based polymorphism support in combination with gson. Given the following classes:

    class ReferencingType{ Base ref;}
    class Base{}
    class A extends Base { String foo;}
    class B extends Base { String bar;}

A serialization of an instance of `ReferencingType` would become:

    {"ref": {
          "a": { "foo": "Hello World"}
    }}

Thus a synthetic object is inserted into the reference, with a property for each sub type (in this case "a"). One way to translate this to TypeScript would be

    interface ReferencingType{ ref: BaseRef}
    interface BaseRef{a: A; b: B}
    interface Base{}
    interface A extends Base {foo: string}
    interface B extends Base{bar: string}

What do you think about supporting this in typescript generator? 
I looked into this on the Java->model side, but adding a bean type without a corresponding Java type seems hard to impossible. Would it make sense to add this on the typescript generation side?

**Comment by vojtechhabarta** [comment]
Firstly I would like to make clear how this feature would look like.

As far as I know GSON doesn't have any "standard" polymorphism support. I saw for example [RuntimeTypeAdapterFactory](https://github.com/google/gson/blob/master/extras/src/main/java/com/google/gson/typeadapters/RuntimeTypeAdapterFactory.java) which uses additional "type" property inside JSON object.

Your case is different but that's ok, typescript-generator could support more ways of polymorphism with GSON. However I wonder how would you design this from users perspective. We could add `gsonConfiguration` parameter (similarly to `jackson2Configuration`) and inside it allow user to specify subclasses. But this has big drawback which is duplication of subclasses info from Java to typescript-generator configuration. Or do you want to use some Java annotations and get subclasses info from user's Java code?

**Comment by ruediste** [comment]
Just to be clear, I stumbled over three polymorphism approaches so far in various places:

1. Use an array. First element: type name, second element: serialized value
1. Property Based. Described above. For example used in the [ElasticSearch query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html) and [Kubernetes](https://kubernetes.io/docs/concepts/storage/volumes/#types-of-volumes)
1. type property

This is all about 2.

I would prefer to get subclass info from user's Java code (annotations). I would start with the following configuration:

**propertyBasedPolymorphismMarker**: fully qualified name of an annotation class. If this annotation is present on a type, references to the type support polymorphism. If nothing else is configured, the simple name of the sub classes, converted from UpperCamel to lowerCamel are used as property names.

**propertyBasePolymorphismNameAnnotation**: fqn of an annotation class. If present on a type, specifies the property name used for this type

**propertyBasePolymorphismNameAnnotationField**: field name of the propertyBasePolymorphismNameAnnotation containing the property name. Defaults to "value".

What do you think?

**Comment by vojtechhabarta** [comment]
Thanks for more detailed description of your intention. I think polymorphism is important feature so it would be good to have support also for this type of polymorphism with Gson.

Just for comparison: in Jackson this can be done using `WRAPPER_OBJECT` mechanism:
``` java
class ReferencingType {
    public Base ref;
}

@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
@JsonSubTypes({
    @JsonSubTypes.Type(A.class),
    @JsonSubTypes.Type(B.class),
})
class Base {
}

@JsonTypeName("a")
class A extends Base {
    public String foo;
}

@JsonTypeName("b")
class B extends Base {
    public String bar;
}
```

Note that there is also `@JsonSubTypes` annotation which lists all possible subtypes. Do you also plan to have something similar?

I can implement support for this form of polymorphism for Jackson first so it would be then easier to implement it for Gson.

But one important difference between Jackson and Gson is that Jackson defines this type of polymorphism (annotations and behavior) but with Gson people can implement it using different types of annotations or even without annotations, have different default property names etc. So I think this feature should be provided as a typescript-generator extensions (instead of in typescript-generator main part) because typescript-generator should not dictate just one way.

One more comment is that disadvantage in Gson case is that this feature assumes there also exists corresponding implementation for Gson so it would be great if you could somehow provide this implementation or at least an example.

BWT if you are designing new API I would rather use polymorphism with "type" (discriminant) property because it is well supported in TypeScript (see https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#tagged-union-types).

**Comment by vojtechhabarta** [comment]
Related pull request (#408) released in version [2.17.558](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.17.558).
